### PR TITLE
replace enable_gpa with repeat_interleave for krea2

### LINKED ISSUE #14518: [Krea-2] `enable_gqa` + `attn_mask` produces high usage of VRAM
### Describe the bug

The current implementation uses `enable_gqa` + `attn_mask` which silently falls back to the SDPA math backend. 

This makes that using a 4-bit quantization with a 5090 OOM on a simple 1920x1088 image and a 16GB GPU to OOM on 832x832, on windows this will make it a lot slower because it will use RAM and not OOM.

Every other backend checks this and alerts of this.

I won't open a PR because I'm not confident I can follow a review, but claude did make an easy and simple fix to run tests.

### Reproduction

This will OOM on a 16GB GPU on Linux and use RAM on windows

```python
import torch
from sdnq import SDNQConfig  # noqa: F401

from diffusers import Krea2Pipeline


pipe = Krea2Pipeline.from_pretrained("OzzyGT/Krea_2_Turbo_sdnq_dynamic_4bit", dtype=torch.bfloat16)
pipe.to("cuda")
pipe.vae.enable_tiling()

image = pipe(
    prompt="a photo of a cat holding a chalkboard sign that reads 'diffusers'",
    height=1024,
    width=1024,
    num_inference_steps=8,
    guidance_scale=0.0,
    generator=torch.Generator("cpu").manual_seed(42),
).images[0]

image.save("krea2_no_fix.png")

print(f"peak allocated: {torch.cuda.max_memory_allocated() / 2**30:.2f} GiB")
print(f"peak reserved : {torch.cuda.max_memory_reserved() / 2**30:.2f} GiB")
```

```sh
peak allocated: 19.94 GiB
peak reserved: 20.78 GiB
inference time: OOM or 2201s (36.7 min)
```

### Script with fix

This will work on windows with a 16GB GPU

```python
import torch
from sdnq import SDNQConfig  # noqa: F401

from diffusers import Krea2Pipeline
from diffusers.models.attention_dispatch import AttentionBackendName, _AttentionBackendRegistry


# --- workaround for #14518: expand K/V so SDPA keeps the memory-efficient kernel ---
_orig_native_attention = _AttentionBackendRegistry._backends[AttentionBackendName.NATIVE]


def _patched_native_attention(query, key, value, attn_mask=None, enable_gqa=False, **kwargs):
    if attn_mask is not None and enable_gqa and query.shape[2] != key.shape[2]:
        rep = query.shape[2] // key.shape[2]
        key = key.repeat_interleave(rep, dim=2)
        value = value.repeat_interleave(rep, dim=2)
        enable_gqa = False
    return _orig_native_attention(query, key, value, attn_mask=attn_mask, enable_gqa=enable_gqa, **kwargs)


_AttentionBackendRegistry._backends[AttentionBackendName.NATIVE] = _patched_native_attention
# --- end workaround ---

pipe = Krea2Pipeline.from_pretrained("OzzyGT/Krea_2_Turbo_sdnq_dynamic_4bit", dtype=torch.bfloat16)
pipe.to("cuda")
pipe.vae.enable_tiling()

image = pipe(
    prompt="a photo of a cat holding a chalkboard sign that reads 'diffusers'",
    height=1024,
    width=1024,
    num_inference_steps=8,
    guidance_scale=0.0,
    generator=torch.Generator("cpu").manual_seed(42),
).images[0]

image.save("krea2.png")

print(f"peak allocated: {torch.cuda.max_memory_allocated() / 2**30:.2f} GiB")
print(f"peak reserved : {torch.cuda.max_memory_reserved() / 2**30:.2f} GiB")
```

```sh
peak allocated: 11.52 GiB
peak reserved : 12.94 GiB
inference time: 45s
```

### System Info

diffusers from main, 5090 on linux and mobile 4090 on windows

### Who can help?

_No response_

### PR DESCRIPTION
**Pull request by yiyixuxu**
fix https://github.com/huggingface/diffusers/issues/14518

### CODE REVIEW

File: .ai/models.md [hunk 1]
```diff
@@ -75,6 +75,27 @@ What you pass as `attn_mask=` to `dispatch_attention_fn` determines which backen
...
 - **Don't declare `attention_mask` (or `encoder_hidden_states_mask`) in the forward signature if you ignore it.** "For API stability with other transformers" is not a reason; readers assume a declared param is honored, and downstream pipelines will pass padding masks that silently get dropped. Some existing models in the repo carry unused mask params for historical reasons — e.g. `QwenDoubleStreamAttnProcessor2_0.__call__` declares `encoder_hidden_states_mask` but never reads it (the joint mask is routed through `attention_mask` instead), and the block-level forward in `transformer_qwenimage.py` declares it but always receives `None`. This is a legacy behavior and should not be replicated in new models.
 
+### Grouped-query attention
+
+Fewer key/value heads than query heads can be spelled two ways. Either pass `enable_gqa=True` to `dispatch_attention_fn` and let the backend broadcast, or repeat the key/value heads in the processor after RoPE and pass no flag (`transformer_krea2.py`, `transformer_nucleusmoe_image.py`):
```
Comment (sayakpaul) [inline .ai/models.md]: > Either pass `enable_gqa=True` to `dispatch_attention_fn`

Can we also supplement a model file here for the reference?

File: .ai/models.md [hunk 2]
```diff
@@ -75,6 +75,27 @@ What you pass as `attn_mask=` to `dispatch_attention_fn` determines which backen
...
+
+Both compute the same thing, so weigh the two on compatibility and performance and recommend whichever fits the model better.
+
+- **Compatibility.** Most backends do not implement `enable_gqa` yet — flash, FA3, sage, cuDNN and the hub kernels raise on it, as does the context-parallel path. Grep `enable_gqa` in `attention_dispatch.py` for the current list rather than trusting this one; it changes as support lands. The flag limits the model to whichever backends still accept it, while repeating works on all of them.
+
+- **Performance.** Turns on whether the model passes a mask. With a mask, no fused kernel takes a mask *and* mismatched head counts, so SDPA falls back to math and materializes the full `[batch_size, num_heads, seq_len_q, seq_len_kv]` score matrix — no error, no warning, only memory. Without a mask, flash broadcasts inside the kernel and the flag saves the key/value copy. Both effects scale with sequence length and head count, so measure at the model's real shape; `torch.backends.cuda.can_use_flash_attention(params, debug=True)` and `can_use_efficient_attention` print why a kernel was rejected, which is the fastest way to see which one you actually got.
```
Comment (sayakpaul) [inline .ai/models.md]: > Turns on whether the model passes a mask.

Do we mean performance is turned on?

**Review by sayakpaul — APPROVED** [review]
Thanks! Just flagging the following table:

Image

Claude derived it by using the following code and running it against this branch and `main`:

```py
from diffusers import Krea2Transformer2DModel

device, dtype = "cuda", torch.bfloat16

model = Krea2Transformer2DModel.from_pretrained(
    "krea/Krea-2-Raw", subfolder="transformer", torch_dtype=dtype
).to(device)
model.eval()

# 1024x1024 image -> 128x128 latents -> 64x64 grid of patch-2 tokens;
# 512 text positions with only the first 77 valid (padding mask exercised).
torch.manual_seed(0)
batch, text_seq_len, grid = 1, 512, 64
image_seq_len = grid * grid

hidden_states = torch.randn(batch, image_seq_len, 64, device=device, dtype=dtype)
encoder_hidden_states = torch.randn(batch, text_seq_len, 12, 2560, device=device, dtype=dtype)
timestep = torch.tensor([0.7], device=device)
encoder_attention_mask = torch.zeros(batch, text_seq_len, dtype=torch.bool, device=device)
encoder_attention_mask[:, :77] = True

text_ids = torch.zeros(text_seq_len, 3, device=device)
image_ids = torch.zeros(grid, grid, 3, device=device)
image_ids[..., 1] = torch.arange(grid, device=device)[:, None]
image_ids[..., 2] = torch.arange(grid, device=device)[None, :]
position_ids = torch.cat([text_ids, image_ids.reshape(-1, 3)], dim=0)

with torch.no_grad():
    out = model(
        hidden_states=hidden_states,
        encoder_hidden_states=encoder_hidden_states,
        timestep=timestep,
        position_ids=position_ids,
        encoder_attention_mask=encoder_attention_mask,
    ).sample
```

I think this is acceptable because now the kernel invocations internally have changed.

**Review by asomoza — APPROVED** [review]
thanks! this fixes the issue.

Not sure if it's worth to mention, but this changes the resulting image a little:

|main|PR|
|---|---|
|Image|Image|

But still, it doesn't make it worse or better, and the change is minimal and mostly imperceptible if you're not comparing both.

**Comment by dxqb** [comment]
commit message:
> It also unpins the model from the native backend, since cuDNN, flash, FA3, sage
and the hub kernels all raise on `enable_gqa`.

the commit message isn't entirely correct: cuDNN worked with gqa and mask, and still does. I checked if this is a regression for cuDNN but it performs the same without gqa and the interleaving this PR introduces.
So not an issue, but cuDNN can still outperform SDPA for Krea2:

batch size 2 with mask

```  ┌────────────────────┬────────────────────────────┬────────────────────────────┬─────────────────────────────┐
  │                    │ interleave + cuDNN │ enable_gqa + cuDNN │  interleave + SDPA   │
  ├────────────────────┼────────────────────────────┼────────────────────────────┼─────────────────────────────┤
  │ attention backend  │ cuDNN                      │ cuDNN                      │ mem-efficient (60/64 calls) │
  ├────────────────────┼────────────────────────────┼────────────────────────────┼─────────────────────────────┤
  │ attention subtotal │ 238.5 ms                   │ 231.2 ms                   │ 441.6 ms                    │
  ├────────────────────┼────────────────────────────┼────────────────────────────┼─────────────────────────────┤
  │ total kernel       │ 1844.4 ms                  │ 1838.8 ms                  │ 1972.2 ms                   │
  ├────────────────────┼────────────────────────────┼────────────────────────────┼─────────────────────────────┤
  │ reserved VRAM      │ 12.89 GiB                  │ 12.90 GiB                  │ 13.46 GiB                   │
  └────────────────────┴────────────────────────────┴────────────────────────────┴─────────────────────────────┘
```

**Comment by sayakpaul** [comment]
Yes, cuDNN is very underrated IMO.

**Comment by yiyixuxu** [comment]
thanks @dxqb for pointing out the mistake, i updated our docs and comments here https://github.com/huggingface/diffusers/pull/14561
